### PR TITLE
Add distance_from_spawn and distance_from_coordinates block/entity conditions

### PR DESCRIPTION
**Pull request by Alluysl**
Add the `distance_from_spawn` and `distance_from_coordinates` entity and block conditions. They're the same but with different default values.

## Fields

The 17 fields are as follows, only 2 are mandatory:

Field  | Type | Default | Description
-------|:----:|:-------:|-------------
`reference` | String | `world_origin` (DFC)<br>`world_spawn` (DFS) |  The point to compare the distance to with.
`offset_x`/`x` | Double | 0.0 | By how much the reference should be offset on the X axis.
`offset_y`/`y` | Double | 0.0 | By how much the reference should be offset on the Y axis.
`offset_z`/`z` | Double | 0.0 | By how much the reference should be offset on the Z axis.
`ignore_x` | Boolean | `false` | Considers the distance on the X axis to be 0.
`ignore_y` | Boolean | `false` | Considers the distance on the Y axis to be 0.
`ignore_z` | Boolean | `false` | Considers the distance on the Z axis to be 0.
`shape` | Shape | `cube` |  See notes below 
`scale_reference_to_dimension` | Boolean | `true` | Whether or not to test for the coordinate the reference would have had if it "took a nether portal" to its new coordinates.
`scale_distance_to_dimension` | Boolean | `false` | Whether or not to scale the calculated distance based on the scale of the dimension the entity/block is in.<br>e.g. multiplies the distance by 8 in the nether, such as if this flag is true, it checks for the distance you'd have to travel in the overworld from a portal at the block/entity to a portal at the reference point.
`comparison` | Comparison | | How the calculated distance should be compared to the specified value.
`compare_to` | Double | | The value to compare the Y position of the block to.
`result_on_wrong_dimension` | Boolean | *optional* | If set, will override the result of the comparison if the entity/block being tested is not in the reference's dimension.
`round_to_digit` | Integer | *optional* | If set, rounds the result to the closest number with the specified amount of digits after the comma. Negative numbers work (e.g. `-2` rounds to multiples of 100).

## Rationale and explanations

- `reference` is future-proofed to add `player_spawn` and `player_natural_spawn` once the `modify_player_spawn` power type is fixed. Note that given the current two references, the scale of the reference's dimension is supposed to be 1.0 since it's the overworld in both cases. If player/power spawns are implemented, this condition will need to be reworked a little (some scaling to do in places, and I imagine two new fields), though I'm of the belief that it won't require breaking backward compatibility.
- `offset_<axis>` and `<axis>` fields add up to each other (used as a cheap method of aliasing) to move the reference point (mostly useful with `world_origin`). The difference is that semantically, for DFS we deal with an offset, and for DFC while it still technically is an offset from `0 0 0` it's more meaningful to think of the coordinates directly (example below). Thus the aliases. My idea would be to only mention the "offsetful" syntax in the documentation for `distance_from_spawn`, and only the "offsetless" syntax in that of `distance_from_coordinates`. By the way, since the center of a block is at `(floor(x) + 0.5, floor(y) + 0.5, floor(z) + 0.5)`, these fields might have to be used even if only to set all three to 0.5. I feel like it's not the job of this condition to do this offset, otherwise it would sure match with people's instinct, but not with f3.
- `ignore_<axis>` can be used for testing the distance in 1D/2D; setting this and the one for Z to `true` makes the condition equal to `height` outside of edge cases (dimensions with a coordinate scale of 0 or wrong shape or can't get World from WorldView).

## Diagram

distance_from_coordinates_scaling

## Example/test powers

Check out the powers added to `testdata` too.

Generic demonstrative power (different from the one in `testdata`):

```json
{
	"type": "apoli:action_over_time",
	"interval": 1,

	"rising_action": {
		"type": "apoli:execute_command",
		"command": "tellraw @s \"You are now in a radius of 8 blocks around 256 64 32\""
	},

	"falling_action": {
		"type": "apoli:execute_command",
		"command": "tellraw @s \"You are now outside the 8-block radius around 256 64 32\""
	},

	"condition": {
		"type": "apoli:distance_from_coordinates",
		"x": 256,
		"y": 64,
		"z": 32,
		"shape": "sphere",
		"comparison": "<",
		"compare_to": 8
	}
}
```

Test power (different from the one in `testdata`):
```json
{
	"type": "apoli:action_over_time",
	"interval": 1,
	"entity_action": {
		"type": "apoli:if_else",

		"if_action": {
			"type": "apoli:execute_command",
			"command": "setblock ~ 255 ~ yellow_stained_glass keep"
		},
		"else_action": {
			"type": "apoli:execute_command",
			"command": "setblock ~ 255 ~ light_blue_stained_glass keep"
		},
		"condition": {
			"type": "apoli:distance_from_spawn",
			"shape": "cube",
			"ignore_y": true,
			"comparison": "<",
			"compare_to": 4
		}
	}
}
```
Result (the low FPS is because I was testing on capped FPS):
Overworld
Nether

## Points of concern and more

- I wish I could split up the `testCondition` method some more, but it appears non-trivial given the edge cases that warrant early returns - else the code would be more complex or unnecessarily allocate objects.
- Some field names seem too long.
- I feel like the "warn once" feature could be useful elsewhere if separated from this class to make it more generic.

### CODE REVIEW

**Comment by Alluysl** [comment]
I'm aware it's probably better to register the `shape` field as an actual shape and not a string. If you're willing to translate that (either way check out the bullet point), feel free to edit the branch, else I'll take a look myself at some point.

**Comment by MerchantPug** [comment]
holy shit

File: src/main/java/io/github/apace100/apoli/power/factory/condition/DistanceFromCoordinatesConditionRegistry.java [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+     * Returns an array of aliases for the condition.
+     * */
+    private static String[] getAliases(){
+        return new String[]{"distance_from_spawn", "distance_from_coordinates"};
+    }
+
+    private static SerializableData getSerializableData(String alias){
+        // Using doubles and not ints because the player position is a vector of doubles and the sqrt function (for the distance) returns a double so we might as well use that precision
+        return new SerializableData()
+            .addFunctionedDefault("reference", SerializableDataTypes.STRING, data -> alias.equals("distance_from_coordinates") ? "world_origin" : "world_spawn") // the reference point
```
Comment (apace100) [inline src/main/java/io/github/apace100/apoli/power/factory/condition/DistanceFromCoordinatesConditionRegistry.java:62]: No need to use `addFunctionedDefault` here - the `alias` is a String provided to the method and therefore "constant". You only need to check it once and then provide a different default string value to the regular `add` method based on what it is.
`addFunctionedDefault` is for when you want a different default value based on the values of other fields provided by the user.

**Comment by apace100** [comment]
Apart from my previous comment, I might replace the `offset_x`, ... and `x`, ... parameters with a vector data type in the future (maybe before release?). But that is nothing for you to worry about right now! :D

Other than that great PR. :) Thanks for putting in the effort of commenting it thoroughly and providing test powers. :) Will help lots in keeping this up-to-date with any changes!
